nat/src/main/java/org/hyperledger/besu/nat/core/AbstractNatManager.java

```diff
@@ -0,0 +1,127 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public abstract class AbstractNatManager implements NatManager {
+  protected static final Logger LOG = LogManager.getLogger();
+
+  protected final NatMethod natMethod;
+
+  protected final AtomicBoolean started = new AtomicBoolean();
+
+  protected AbstractNatManager(final NatMethod natMethod) {
+    this.natMethod = natMethod;
+  }
+
+  protected abstract void doStart();
+
+  protected abstract void doStop();
+
+  protected abstract CompletableFuture<String> retrieveExternalIPAddress();
+
+  @Override
+  public NatMethod getNatMethod() {
+    return natMethod;
+  }
+
+  @Override
+  public boolean isStarted() {
+    return started.get();
+  }
+
+  @Override
+  public CompletableFuture<String> queryExternalIPAddress() {
+    if (!isStarted()) {
+      throw new IllegalStateException("Cannot call queryExternalIPAddress() when in stopped state");
+    }
+    return retrieveExternalIPAddress();
+  }
+
+  @Override
+  public CompletableFuture<String> queryLocalIPAddress() {
+    final CompletableFuture<String> future = new CompletableFuture<>();
+    Executors.newCachedThreadPool()
+        .submit(
+            () -> {
+              try {
+                future.complete(InetAddress.getLocalHost().getHostAddress());
+              } catch (UnknownHostException e) {
+                future.completeExceptionally(e);
+              }
+            });
+    return future;
+  }
+
+  @Override
+  public void start() {
+    if (started.compareAndSet(false, true)) {
+      doStart();
+    } else {
+      LOG.warn("Attempt to start an already-started {}", getClass().getSimpleName());
+    }
+  }
+
+  @Override
+  public void stop() {
+    if (started.compareAndSet(true, false)) {
+      doStop();
+    } else {
+      LOG.warn("Attempt to stop an already-stopped {}", getClass().getSimpleName());
+    }
+  }
+
+  @Override
+  public NatPortMapping getPortMapping(
+      final NatServiceType serviceType, final NetworkProtocol networkProtocol) {
+    try {
+      final List<NatPortMapping> natPortMappings =
+          getPortMappings().get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+      final Optional<NatPortMapping> foundPortMapping =
+          natPortMappings.stream()
+              .filter(
+                  c ->
+                      c.getNatServiceType().equals(serviceType)
+                          && c.getProtocol().equals(networkProtocol))
+              .findFirst();
+      return foundPortMapping.orElseThrow();
+    } catch (NoSuchElementException e) {
+      throw new IllegalArgumentException(
+          String.format("Required service type not found : %s %s", serviceType, networkProtocol));
```

Comment:
Can this `IllegalArgumentException` be put into the `orElseThrow`?

Reply:
yes you're right I just made the change